subsys/net/ip/connection.c

```diff
@@ -598,8 +616,42 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 			}
 
 			if (best_rank < NET_CONN_RANK(conn->flags)) {
-				best_rank = NET_CONN_RANK(conn->flags);
-				best_match = conn;
+				struct net_pkt *mcast_pkt;
+
+				if (!is_mcast_pkt) {
+					best_rank = NET_CONN_RANK(conn->flags);
+					best_match = conn;
+
+					continue;
+				}
+
+				/* If we have a multicast packet, and we found
+				 * a match, then deliver the packet immediately
+				 * to the handler. As there might be several
+				 * sockets interested about these, we need to
+				 * clone the received pkt.
+				 */
+
+				NET_DBG("[%p] mcast match found cb %p ud %p",
+					conn, conn->cb,	conn->user_data);
+
+				mcast_pkt = net_pkt_clone(pkt, CLONE_TIMEOUT);
```

Comment:
Do you think we could do a shallow clone here?
It would save some memory but it's also risky that a receiver would change something in the pkt.

Reply:
As we cannot guarantee that the apps won't modify the data, the net_buf needs to be cloned too.